src/Microsoft.ML.OnnxTransform/OnnxTransform.cs

```diff
@@ -333,91 +352,103 @@ protected override Delegate MakeGetter(IRow input, int iinfo, Func<int, bool> ac
                 var activeOutputColNames = _parent.Outputs.Where((x, i) => activeOutput(i)).ToArray();
                 var type = OnnxUtils.OnnxToMlNetType(_parent.Model.ModelInfo.OutputsInfo[iinfo].Type).RawType;
                 Host.Assert(type == _parent.OutputTypes[iinfo].ItemType.RawType);
-                var srcTensorGetters = GetTensorValueGetters(input, _inputColIndices, _isInputVector, _inputOnnxTypes, _inputTensorShapes);
-                return Utils.MarshalInvoke(MakeGetter<int>, type, input, iinfo, srcTensorGetters, activeOutputColNames, outputCache);
+                var srcNamedValueGetters = GetNamedOnnxValueGetters(input, _parent.Inputs, _inputColIndices, _isInputVector, _inputOnnxTypes, _inputTensorShapes);
+                return Utils.MarshalInvoke(MakeGetter<int>, type, input, iinfo, srcNamedValueGetters, activeOutputColNames, outputCache);
             }
 
-            private Delegate MakeGetter<T>(IRow input, int iinfo, ITensorValueGetter[] srcTensorGetters, string[] activeOutputColNames, OutputCache outputCache)
+            private Delegate MakeGetter<T>(IRow input, int iinfo, INamedOnnxValueGetter[] srcNamedValueGetters, string[] activeOutputColNames, OutputCache outputCache)
             {
                 Host.AssertValue(input);
                 ValueGetter<VBuffer<T>> valuegetter = (ref VBuffer<T> dst) =>
                 {
-                    UpdateCacheIfNeeded(input.Position, srcTensorGetters, activeOutputColNames, outputCache);
-                    var tensor = outputCache.Outputs[_parent.Outputs[iinfo]];
-                    var editor = VBufferEditor.Create(ref dst, tensor.GetSize());
-                    OnnxUtils.CopyTo(tensor, editor.Values);
+                    UpdateCacheIfNeeded(input.Position, srcNamedValueGetters, activeOutputColNames, outputCache);
+                    var namedOnnxValue = outputCache.Outputs[_parent.Outputs[iinfo]];
+                    var editor = VBufferEditor.Create(ref dst, namedOnnxValue.AsTensor<T>().Count());
+                    var denseTensor = namedOnnxValue.AsTensor<T>() as System.Numerics.Tensors.DenseTensor<T>;
```

Comment:
You should switch these two lines around so you don't need to do this call twice in a row. Also, don't use the `Count()` method, use `.Length` instead.

```C#
var denseTensor = namedOnnxValue.AsTensor<T>() as System.Numerics.Tensors.DenseTensor<T>;
if (denseTensor == null)
    throw Host.Except($"Output column {namedOnnxValue.Name} doesn't contain a DenseTensor of expected type {typeof(T)}");
var editor = VBufferEditor.Create(ref dst, denseTensor.Length);
```

Reply:
Done. Much cleaner this way.

---
In reply to: [236335318](https://github.com/dotnet/machinelearning/pull/1717#discussion_r236335318) [](ancestors = 236335318)